mesonbuild/msetup.py

```diff
@@ -251,7 +251,11 @@ def _generate(self,
             if self.options.profile:
                 fname = f'profile-{intr.backend.name}-backend.log'
                 fname = os.path.join(self.build_dir, 'meson-logs', fname)
-                profile.runctx('intr.backend.generate()', globals(), locals(), filename=fname)
+                profile.runctx('gen_result = intr.backend.generate('
+                               'capture=capture,'
+                               'captured_compile_args_per_buildtype_and_target=captured_compile_args_per_buildtype_and_target)', globals(), locals(), filename=fname)
+                captured_compile_args = locals()['gen_result']
+                assert ((captured_compile_args is None) or isinstance(captured_compile_args, dict))
```

Comment:
This is not a real assert, which may explain why mypy didn't do type narrowing on it.

Reply:
That's incredibly helpful.
It looks like an assert, it parses/interprets without any complaint, and it also behaves correctly -
```
>>> assert (({} is None) or isinstance({}, dict))
>>> assert ((None is None) or isinstance(None, dict))
>>> assert ((123 is None) or isinstance(123, dict))
... AssertionError
```
So would you care to share your insight on what makes this not a real assert?

Reply:
Let's rewrite this assert to be semantically identical, but more easily readable.

```python
isnone: bool = captured_compile_args is None
isdict: bool = isinstance(captured_compile_args, dict)
test_statement: T.Tuple[bool, bool] = (isnone, isdict)

assert test_statement
```

This doesn't generate the same bytecode, but the only difference is that it needs to STORE_NAME before the assert.

Reply:
We have quite different definitions of easily readable 😄 

Even with the those contortions to seemingly try to spoon-feed mypy, it still thinks we'll can end up returning an Any type... But I suppose I (and now you also) must be misunderstanding the assert and/or not providing sufficient information to mypy to deduce narrow the type with the assert. 

Also, you haven't explained why this is not a real assert.  As far as I can see, everything about it does what it should and that all signs point to a limitation with mypy's type deduction , but I'd still love to learn what makes it an incorrect or not real assert.

Reply:
[054e714](https://github.com/mesonbuild/meson/pull/11934/commits/054e714bf19b9f54aaa9ad86f29e72ba564441ac) seems to be enough of a dance, to get mypy to figure things out.